### PR TITLE
 Share Dashboard cookie-based session with Pegasus

### PR DESCRIPTION
**Pull request by wjordan**
This PR is an attempt at refactoring some of the Code Studio session-cookie sharing in the Pegasus (Sinatra) stack. The goal is to make our use of Code Studio sessions in parts of the Pegasus stack a bit more flexible and maintainable moving forward.

With this PR I wanted to get rid of some hard-to-understand code, and to avoid minor implementation changes to session-cookie encryption in Rails version updates from breaking this parallel decryption routine again in the future.

A longer-term sub-goal is to start incorporating some additional Rails components into the Pegasus/Sinatra stack where appropriate, and where it can potentially simplify our code, eliminate parallel feature implementations, etc., with an eye towards eventually having a more unified (or at least more standard/consistent) web-application stack underlying both.

A side benefit is that the full unencrypted Code Studio session data will now be available in the standard `rack.session` env variable throughout Pegasus (via e.g., [`request.session`](https://github.com/rack/rack/blob/2.0.4/lib/rack/request.rb#L144)), which might make it easier to leverage Rack-middleware gems or implement other features needing access to other pieces of Code Studio session-state more easily. (Note however that for public-cached pages which includes most `code.org/` routes by default the session state is still unavailable by design, so anything session-related only applies to pages where the session-cookie is whitelisted in the HTTP cache config.)

#### Implementation
The existing `ActionDispatch::Session::CookieStore` middleware included in Dashboard (Rails) to manage session-cookie state is now also included in the Pegasus stack, replacing the parallel implementation in `user_id_from_session_cookie` and its hard-coded magic-strings and constants (which was manually kept in sync with Rails' internal implementation). In order to supply the environment-config defaults Rails uses, I added a `DummyApp < Rails::Application` class to populate the [`Rails.application.env_config`](https://github.com/rails/rails/blob/v5.0.6/railties/lib/rails/application.rb#L242-L266) singleton with some necessary [defaults](https://github.com/rails/rails/blame/v5.0.6/actionpack/lib/action_dispatch/railtie.rb#L15-L18), and added a `RailsEnvConfig` middleware to the Pegasus stack that injects those values into the rack environment so they can be used by the `CookieStore`.

This implementation is also a bit complex and makes some non-public-API assumptions about Rails internals in its own way, so some of the benefit definitely washes out. I think it makes at least some slight improvements to the overall balance though.

### CODE REVIEW

**Review by islemaster — COMMENTED** [review]
As far as I understand it, this does seem to leave us in a better place and be more future-proof than the current solution.  Is there any other complex code dealing with cookies that falls out after this change?

Is there a testing story for code like this?  Session management is one of those things that we'll notice right away if it's broken, via our UI tests at least, but I don't fully understand the matrix here and wouldn't be surprised if there are some edge cases with sessions that our automated tests don't capture.

From a user point of view does anything change? e.g. are any existing cookies invalidated when this ships (I think no, but see that `CDO.sinatra_session_secret` goes away, so wondering what the ramifications of that are).

File: lib/cdo/rails_env_config.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Load a minimal dummy Rails app for env defaults if one isn't already loaded.
+# This provides
+unless defined?(Rails) && Rails.application
+  # Skip Sass-plugin railtie because it depends on ActionController.
+  Sass::RAILS_LOADED = true
```
Comment (islemaster) [inline lib/cdo/rails_env_config.rb:5]: Is this a list of exceptions that's likely to grow over time as our application gets more complicated?  I started one of those here a year ago and... well, I'm not thrilled about it:

https://github.com/code-dot-org/code-dot-org/blob/a6707d00127e579cd9b702160a58c7616aee689a/apps/webpack.js#L9-L21

Comment (wjordan) [reply lib/cdo/rails_env_config.rb:5]: I hope this list doesn't grow- I actually consider this to be a missing-require 'bug' in the Sass-Rails Railtie (though it's definitely rare for a `Rails::Application` to be created without `ActionController::Base` already loaded, I don't think the Rails API forbids it). I could file a fix upstream and track the issue, but for now this was a quick workaround.

**Comment by wjordan** [comment]
Thanks, your questions hit the exact things I'd like to focus on before merging / considering this work done.

> Is there a testing story for code like this?

Nothing yet. This code never had any tests around it, which is partly why it remained such a fragile and opaque integration piece. I'm just starting to think about the right approach to adding meaningful coverage around it.

> From a user point of view does anything change? e.g. are any existing cookies invalidated when this ships (I think no, but see that CDO.sinatra_session_secret goes away, so wondering what the ramifications of that are).

Existing cookies shouldn't be invalidated. Once I figure out a testing plan I want to verify that extra cookies aren't being unnecessarily written in responses, and that the [`Rack::Protection`](https://github.com/sinatra/sinatra/tree/v2.0.1/rack-protection) middleware (injected into the Pegasus stack by Sinatra's [default middleware](https://github.com/sinatra/sinatra/blob/v2.0.1/lib/sinatra/base.rb#L1643-L1651) list) isn't going to start [dropping sessions](https://github.com/sinatra/sinatra/blob/v2.0.1/lib/sinatra/base.rb#L1693) unexpectedly if we don't have everything configured as we should in all environments. (This is the reason f41aa91 was necessary in this PR.)

**Comment by wjordan** [comment]
Also, the Sinatra-specific rack-session seems to be used by a single endpoint (#7509, more endpoints added in #7688 but was reverted): https://github.com/code-dot-org/code-dot-org/blob/a6707d00127e579cd9b702160a58c7616aee689a/pegasus/config.ru#L3-L5

**Pull request closed without merging**